File: Array/easy/find_the_union.cpp
//
// Created by santosh on 12/10/25.
//

#include<bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<int> unionByMap(vector<int>& m, vector<int>& n)
    {
        map<int, int> freq;
        vector<int> Union;
        for (auto element : m)
        {
            freq[element]++;
        }
        for (auto element : n)
        {
            freq[element]++;
        }

        for (auto element : freq)
        {
            Union.push_back(element.first);
        }

        return Union;
    }

    vector<int> unionBySet(vector<int>& m, vector<int>& n)
    {
        set<int> res;
        vector<int> Union;
        for (auto element : m)
        {
            res.insert(element);
        }
        for (auto element : n)
        {
            res.insert(element);
        }
        for (auto element : res)
        {
            Union.push_back(element);
        }

        return Union;
    }

    vector<int> unionBy2Pointers(const vector<int>& arr1, const vector<int>& arr2)
    {
        vector<int> Union;
        int i = 0, j = 0;

        while (i < arr1.size() && j < arr2.size())
        {
            if (arr1[i] < arr2[j])
            {
                if (Union.empty() || Union.back() != arr1[i])
                {
                    Union.push_back(arr1[i]);
                }
                i++;

            }
            else if (arr1[i] > arr2[j])
            {
                if (Union.empty() || Union.back() != arr2[j])
                {
                    Union.push_back(arr2[j]);

                };
                j++;
            }
            else
            {
                Union.push_back(arr1[i]);
                i++;
                j++;
            }
        }
        for (i; i < arr1.size(); i++)
        {
            Union.push_back(arr1[i]);
        }
        for (j; j < arr2.size(); j++)
        {
            Union.push_back(arr2[j]);
        }
        return Union;
    }

};


int main()
{
    vector<int> m = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    vector<int> n = {2, 3, 4, 4, 5, 11, 12};

    Solution s;
    vector<int> mapRes = s.unionByMap(m, n);
    for (auto element : mapRes)
    {
        cout << element << " ";
    }
    cout << endl;
    vector<int> setRes = s.unionBySet(m, n);
    for (auto element : setRes)
    {
        cout << element << " ";
    }
    cout << endl;
    vector<int> twoPointersRes = s.unionBy2Pointers(m, n);
    for (auto element : twoPointersRes)
    {
        cout << element << " ";
    }
}


// Output: {1,2,3,4,5,6,7,8,9,10,11,12}

//These problem can be solved in 3 ways

// MAP
// 1.Create a frequency ordered map and and loop over each array to store the elements in a frequency map it will insert thr elements in a sorted array
// at the end loop over the frequency map and insert into the union array
// TIME COMPLEXITY inserting an element into a ordered map takes O(logn) times
// so here O(m+n)log(m+n)
// space complexity o(max(m,n)) for map and O(m+n) for unique array

// SET
// Insert the elements in a set from both the first array and second array
// insertion of element in a ordered set takes the O(log n) times
// so here the 2 array means O(m+n)log(m+n)
// space complexity o(max(m,n)) for set and O(m+n) for unique array

File: Array/easy/find_the_union.py
# Find the union of two sorted arrays
# Created by santosh

from collections import OrderedDict


class Solution:
    def union_by_map(self, m: list[int], n: list[int]) -> list[int]:
        freq: dict[int, int] = {}
        for el in m:
            freq[el] = freq.get(el, 0) + 1
        for el in n:
            freq[el] = freq.get(el, 0) + 1
        return sorted(freq.keys())

    def union_by_set(self, m: list[int], n: list[int]) -> list[int]:
        res = set(m) | set(n)
        return sorted(res)

    def union_by_2_pointers(self, arr1: list[int], arr2: list[int]) -> list[int]:
        union: list[int] = []
        i, j = 0, 0

        while i < len(arr1) and j < len(arr2):
            if arr1[i] < arr2[j]:
                if not union or union[-1] != arr1[i]:
                    union.append(arr1[i])
                i += 1
            elif arr1[i] > arr2[j]:
                if not union or union[-1] != arr2[j]:
                    union.append(arr2[j])
                j += 1
            else:
                union.append(arr1[i])
                i += 1
                j += 1

        while i < len(arr1):
            union.append(arr1[i])
            i += 1
        while j < len(arr2):
            union.append(arr2[j])
            j += 1

        return union


if __name__ == "__main__":
    m = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    n = [2, 3, 4, 4, 5, 11, 12]

    s = Solution()
    print("Union by map:", s.union_by_map(m, n))
    print("Union by set:", s.union_by_set(m, n))
    print("Union by 2 pointers:", s.union_by_2_pointers(m, n))

# Output: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
